### PR TITLE
[MI-2350][server]: Added a feature to update the status of an already approved/rejected Pipeline release request.

### PR DESCRIPTION
**Pull request by avas27JTG**
Added a feature to update the status of an already approved/rejected Pipeline release request.
image

### CODE REVIEW

**Review by raghavaggarwal2308 — APPROVED** [review]

File: server/plugin/api.go [hunk 1]
```diff
@@ -1170,11 +1170,7 @@ func (p *Plugin) handlePipelineApproveOrRejectRequest(w http.ResponseWriter, r *
 	postActionIntegrationRequest := &model.PostActionIntegrationRequest{}
 	if err := decoder.Decode(&postActionIntegrationRequest); err != nil {
 		// TODO: prevent posting any error message except oAuth in DM for now and use dialog for all such cases
-		if _, DMErr := p.DM(mattermostUserID, constants.GenericErrorMessage, true); DMErr != nil {
-			p.API.LogError("Failed to DM", "Error", DMErr.Error())
-		}
-
-		p.API.LogError("Error decoding PostActionIntegrationRequest params: ", err.Error())
+		p.handlePipelineApprovalRequestUpdateError("Error decoding PostActionIntegrationRequest params: ", mattermostUserID, err)
```
Comment (ayusht2810) [inline server/plugin/api.go:1173]:
```suggestion
		p.handlePipelineApprovalRequestUpdateError("Error decoding PostActionIntegrationRequest param: ", mattermostUserID, err)
```

Comment (ayusht2810) [reply server/plugin/api.go:1173]: are there multiple params, as we are only printing `mattermostUserID`

File: server/plugin/client.go [hunk 2]
```diff
@@ -310,6 +311,20 @@ func (c *client) GetSubscriptionFilterPossibleValues(request *serializers.GetSub
 	return subscriptionFiltersResponse, statusCode, nil
 }
 
+func (c *client) GetApprovalDetails(organization, projectName, mattermostUserID string, approvalID int) (*serializers.PipelineApprovalDetails, int, error) {
+	pipelineApproveRequestURL := fmt.Sprintf(constants.PipelineApproveRequest, organization, projectName, approvalID)
```
Comment (ayusht2810) [inline server/plugin/client.go:315]: can't we use this value directly instead of creating a new variable?

Comment (avas27JTG) [reply server/plugin/client.go:329]: this is to make it more readable

**Review by ayusht2810 — APPROVED** [review]

File: server/plugin/api.go [hunk 3]
```diff
@@ -1188,55 +1184,74 @@ func (p *Plugin) handlePipelineApproveOrRejectRequest(w http.ResponseWriter, r *
...
+		p.handlePipelineApprovalRequestUpdateError(constants.GenericErrorMessage, mattermostUserID, updatePipelineApprovalRequestErr)
+		p.handleError(w, r, &serializers.Error{Code: http.StatusInternalServerError, Message: updatePipelineApprovalRequestErr.Error()})
 		return
 	}
 
 	response := &model.PostActionIntegrationResponse{}
 	p.returnPostActionIntegrationResponse(w, response)
 }
 
+func (p *Plugin) UpdatePipelineReleaseApprovalPost(requestType, postID, mattermostUserID string) error {
```
Comment (manojmalik20) [inline server/plugin/api.go:1226]: This function shouldn't be in `api.go`.

File: server/plugin/api.go [hunk 4]
```diff
@@ -1188,55 +1184,74 @@ func (p *Plugin) handlePipelineApproveOrRejectRequest(w http.ResponseWriter, r *
...
+	model.ParseSlackAttachment(post, []*model.SlackAttachment{slackAttachment})
+	if _, err := p.API.UpdatePost(post); err != nil {
+		p.handlePipelineApprovalRequestUpdateError("Error in updating post", mattermostUserID, err)
+		return err
+	}
+
+	return nil
+}
+
+func (p *Plugin) handlePipelineApprovalRequestUpdateError(errorMessage, mattermostUserID string, err error) {
```
Comment (manojmalik20) [inline server/plugin/api.go:1248]: Same.

File: server/plugin/api.go [hunk 5]
```diff
@@ -1188,55 +1184,74 @@ func (p *Plugin) handlePipelineApproveOrRejectRequest(w http.ResponseWriter, r *
...
+		if updatePipelineReleaseApprovalPostErr := p.UpdatePipelineReleaseApprovalPost(requestType, postActionIntegrationRequest.PostId, mattermostUserID); updatePipelineReleaseApprovalPostErr != nil {
+			p.handlePipelineApprovalRequestUpdateError(constants.GenericErrorMessage, mattermostUserID, updatePipelineReleaseApprovalPostErr)
+			p.handleError(w, r, &serializers.Error{Code: http.StatusInternalServerError, Message: updatePipelineReleaseApprovalPostErr.Error()})
 			return
 		}
-
 	case http.StatusBadRequest:
-		if _, DMErr := p.DM(mattermostUserID, fmt.Sprintf(constants.ErrorUpdatingNonPendingPipelineRequest, approvalID), true); DMErr != nil {
-			p.API.LogError("Failed to DM", "Error", DMErr.Error())
+		pipelineApprovalDetails, statusCode, getApprovalDetailsErr := p.Client.GetApprovalDetails(organization, projectName, mattermostUserID, approvalID)
```
Comment (manojmalik20) [inline server/plugin/api.go:1196]:
```suggestion
		pipelineApprovalDetails, statusCode, err := p.Client.GetApprovalDetails(organization, projectName, mattermostUserID, approvalID)
```
IMO, using such big names for errors here is reducing the simplicity and readability of the code here.

Comment (avas27JTG) [reply server/plugin/api.go:1200]: Yes it is for readability and also to prevent chances of shadow declaration with variable name `err`

**Review by manojmalik20 — APPROVED** [review]

File: server/plugin/api.go (hunk 5 above)
Comment (manojmalik20) [reply server/plugin/api.go:1200]: I think you used such big names for increasing readability but IMO, it is decreasing the readability and simplicity of the code here. We can always avoid shadow declaration of `err` by using linter but we should keep the code as simple as possible.

**Pull request merged**